[polly] Add profitability check for expanded region.

huihzhang opened the pull request:
Region expansion may append a basic block that contains memory accesses not used in loops of original region, or any loops in regions to be joined. These additional memory accesses add complexity in building Scop, runtime alias checks and compute for optimization schedule. In certain cases may make original region unfeasible to analyze, e.g., access to a memory location that was loop invariant for loop in the origional region.

This patch adds a profitability check for region expansion. Mark the expanded region as unprofitable, if it includes basic blocks with memory accesses not used in loops of expanded region.

huihzhang committed:
[polly] Add profitability check for expanded region.

Region expansion may append a basic block that contains memory accesses not
used in loops of original region, or any loops in regions to be joined. These
additional memory accesses add complexity in building Scop, runtime alias checks
and compute for optimization schedule. In certain cases may make original region
unfeasible to analyze, e.g., access to a memory location that was loop invariant
for loop in the origional region.

This patch adds a profitability check for region expansion. Mark the expanded
region as unprofitable, if it includes basic blocks with memory accesses not
used in loops of expanded region.

huihzhang commented:
Ping @Meinersbur @efriedma-quic 
If there are questions or adjustments needed, please do let me know?

huihzhang commented:
Ping ?

huihzhang commented:
Ping @Meinersbur @efriedma-quic ?

Meinersbur commented on polly/lib/Analysis/ScopDetection.cpp:
```diff
@@ -1621,6 +1626,12 @@ void ScopDetection::findScops(Region &R) {
 
+    if (!isRegionExpansionProfitable(*ExpandedR, LI)) {
+      removeCachedResults(*ExpandedR);
+      delete ExpandedR;
```
I think this should be moved inside `expandRegion` such that it returns nullptr when this test fails. Also avoids the manual memory management.

huihzhang replied:
Actually, region expansion profitability check should be done at the completion of `expandRegion`.

Because expanding a region to merge adjacent loops may involve multiple steps: 1) expand to include the exit block of current region, this added block may happen to be the preheader block of the next loop to be merged; 2) expand again to include all the blocks of the next loop; 3) repeat step1 and step2, or add other blocks from given CFG.

During the first expansion, the extra block added does not belong to any loops of the expandedRegion, but is actually the preheader block of the adjacent loop to be merged. Then the loop invariant loads for the next loop is going to cause the very first expansion to fail. Eventually region expansion will fail to merge with the adjacent loop, which should have been merged at the second expansion. 

test/CodeGen/region_expansion_profitability_check_loopfusion_1.ll added can be used to reproduce this issue.

Meinersbur replied:
This passes all the tests (after fixing the 3 new ones):
```c++
Region *ScopDetection::expandRegion(Region &R) {
  ...
  while (ExpandedRegion) {
     ...
  }

  if (LastValidRegion && !isRegionExpansionProfitable(*LastValidRegion, LI)) {
    POLLY_DEBUG(dbgs() << "Non-profitable region: "
                       << LastValidRegion->getNameStr() << "\n");
    removeCachedResults(*LastValidRegion.get());
    return nullptr;
  }

  POLLY_DEBUG({
    if (LastValidRegion)
      dbgs() << "\tto " << LastValidRegion->getNameStr() << "\n";
    else
      dbgs() << "\tExpanding " << R.getNameStr() << " failed\n";
  });

  return LastValidRegion.release();
}
```

Meinersbur commented on polly/lib/Analysis/ScopDetection.cpp:
```diff
@@ -1749,6 +1760,56 @@ bool ScopDetection::isProfitableRegion(DetectionContext &Context) const {
+
+  // Collect outermost loops from expanded region.
+  SmallPtrSet<const Loop *, 2> Loops;
+  for (auto BB : ExpandedRegion.blocks()) {
```
```suggestion
  for (BasicBlock *BB : ExpandedRegion.blocks()) {
```

Meinersbur commented on polly/lib/Analysis/ScopDetection.cpp:
```diff
@@ -1749,6 +1760,56 @@ bool ScopDetection::isProfitableRegion(DetectionContext &Context) const {
+      Loops.insert(L);
+  }
+
+  if (Loops.size() == 0) {
```
```suggestion
  if (Loops.empty()) {
```

Meinersbur commented on polly/lib/Analysis/ScopDetection.cpp:
```diff
@@ -1749,6 +1760,56 @@ bool ScopDetection::isProfitableRegion(DetectionContext &Context) const {
+
+  // Return region expansion as unprofitable, if it contains basic blocks with
+  // memory accesses not used in outermost loops of the expanded region.
+  for (auto BB : ExpandedRegion.blocks()) {
```
```suggestion
  for (BasicBlock *BB : ExpandedRegion.blocks()) {
```

Meinersbur commented on polly/lib/Analysis/ScopDetection.cpp:
```diff
@@ -1749,6 +1760,56 @@ bool ScopDetection::isProfitableRegion(DetectionContext &Context) const {
+                     << ExpandedRegion.getNameStr() << "\n");
+
+  // Collect outermost loops from expanded region.
+  SmallPtrSet<const Loop *, 2> Loops;
```
```suggestion
  SmallPtrSet<const Loop *, 2> OutermostLoops;
```
Suggestion for a more descriptive name

Meinersbur commented on polly/lib/Analysis/ScopDetection.cpp:
```diff
@@ -1749,6 +1760,56 @@ bool ScopDetection::isProfitableRegion(DetectionContext &Context) const {
+    if (BB == ExpandedRegion.getEntry())
+      continue;
+
+    // Skip loop preheader block that may contain loop invariant loads.
```
```suggestion
    // Only consider the expansion blocks added in addition to the loops. Also ignore preheader because those contains loop invariant loads.
```

Meinersbur commented on polly/include/polly/ScopDetection.h:
```diff
@@ -345,6 +345,20 @@ class ScopDetection {
 
+  /// Check if an expanded region is profitable to optimize.
+  ///
+  /// An expanded region may include basic blocks with memory accesses that
+  /// are not used in loops of the expanded region. These memory accesses add
```
The actual test looks for any memory access, not just accesses not used in the loop.

huihzhang replied:
I have updated the checks to look for memory accesses that are not used in loops.

Meinersbur commented on polly/lib/Analysis/ScopDetection.cpp:
```diff
@@ -1749,6 +1760,56 @@ bool ScopDetection::isProfitableRegion(DetectionContext &Context) const {
+      continue;
+
+    if (llvm::any_of(*BB, [](const Instruction &I) {
+          return isa<LoadInst>(&I) || isa<StoreInst>(&I);
```
There are other memory access instructions such as `memset`. I suggest to use `I.mayReadOrWriteMemory()` instead.

Meinersbur commented on polly/test/ScopInfo/user_provided_assumptions.ll:
```diff
@@ -4,6 +4,12 @@
 ; CHECK:      remark: <unknown>:0:0: SCoP begins here.
 ; CHECK-NEXT: remark: <unknown>:0:0: Use user assumption: [M, N] -> {  : N <= 2147483647 - M }
 ; CHECK-NEXT: remark: <unknown>:0:0: Use user assumption: [M, N] -> {  : -2147483648 - M <= N <= 2147483647 - M }
+; CHECK-NEXT: remark: <unknown>:0:0: Use user assumption: [M, N] -> {  : 0 < M <= 100 and -2147483648 - M <= N <= 2147483647 - M }
```
Where do the additional remarks come from?

It would be great to have a diagnostic remark for the reason as well. See `ScopDetectionDiagnostic.h`

huihzhang replied:
The first heuristic I posted consider region expansion as unprofitable. the 'if' block and loop are treated as separate regions to build scop.
```
 if (N + M == -1)
        C[0][0] = 0;
 // Loop following
```

Now, the updated heuristic looks for additional loops following unprofitable blocks to encourage loop fusion. Region expansion is successful, and restore original behavior.

Meinersbur reviewed (commented):
This currently checks for any memory accesses in the expanded blocks, not just unrelated ones. 

Even if it was checking that the expanded blocks only contain unrelated accesses, I wonder whether it is too conservative. E.g.

```c
for (int i = 0; i < n; ++i) 
  B[i] = A[i];
*unrelated=42;
for (int j = 0; j < n; ++j) 
  C[j] = B[j];
```
I think we still want to fuse the loop even though there is an unrelated access in-between.

huihzhang committed:
Update region expansion profitability check.

1. When checking unrelated memory accesses of extra blocks, check exactly for
memory accesses that are not used in any loop of the expanded region, instead
of checking for any memory accesses.
2. Encourage loop fusion if extra block contains unrelated memory accesses, but
is followed by another loop. Accept such region expansion.
3. Address review commments.

huihzhang commented:
Thanks a lot @Meinersbur for the review feedbacks!

I have updated the region expansion heuristic, and addressed the following issues:
1) Check exactly for memory accesses that are not used in any loop of the expanded region;
2) Accept region expansion if unprofitable blocks added are followed by another loop, to encourage loop fusion.
A similar test with unrelated memory accesses in between loops is added into test/CodeGen/region_expansion_profitability_check_loopfusion_2.ll
The updated heuristic is accepting such case, and fusion is successful.

Please help take a look of my update, let me know what you think ? Thanks a lot!

Meinersbur commented on polly/test/CodeGen/region_expansion_profitability_check.ll:
```diff
@@ -0,0 +1,138 @@
+; RUN: opt -polly-use-llvm-names -polly-region-expansion-profitability-check=1 \
```
```suggestion
; RUN: opt  %loadNPMPolly -polly-use-llvm-names -polly-region-expansion-profitability-check=1 \
```

Meinersbur commented on polly/test/CodeGen/region_expansion_profitability_check_loopfusion_1.ll:
```diff
@@ -0,0 +1,56 @@
+; RUN: opt -polly-use-llvm-names -polly-region-expansion-profitability-check=1 \
```
```suggestion
; RUN: opt  %loadNPMPolly -polly-use-llvm-names -polly-region-expansion-profitability-check=1 \
```

Meinersbur commented on polly/test/CodeGen/region_expansion_profitability_check_loopfusion_2.ll:
```diff
@@ -0,0 +1,70 @@
+; RUN: opt -polly-use-llvm-names -polly-region-expansion-profitability-check=1 \
```
```suggestion
; RUN: opt  %loadNPMPolly -polly-use-llvm-names -polly-region-expansion-profitability-check=1 \
```

Meinersbur commented on polly/test/CodeGen/region_expansion_profitability_check.ll:
```diff
@@ -0,0 +1,138 @@
+; RUN: opt -polly-use-llvm-names -polly-region-expansion-profitability-check=1 \
+; RUN: "-passes=scop(polly-opt-isl,print<polly-ast>)" -disable-output < %s | \
+; RUN: FileCheck %s --check-prefix=HEURISTIC
+; RUN: opt -polly-region-expansion-profitability-check=0 -passes=polly-codegen \
```
```suggestion
; RUN: opt %loadNPMPolly -polly-use-llvm-names -polly-region-expansion-profitability-check=0 -passes=polly-codegen \
```

Meinersbur commented on polly/lib/Analysis/ScopDetection.cpp:
```diff
@@ -1749,6 +1760,84 @@ bool ScopDetection::isProfitableRegion(DetectionContext &Context) const {
+        llvm::any_of(*BB, [&](const Instruction &I) {
+          if (!I.mayReadOrWriteMemory())
+            return false;
+          if (I.user_empty())
+            return false;
```
Isn't a store's user list always empty? Wouldn't you look for the definition of the value it stores in that case? If the user list is empty, the following loop would have no iterations anyway. 

Also, what would matter is whether the accesses is to the same memory region as something in the loop, which would complicate the dependence and array subscript analysis. Not the value that us being loaded (unless you specifically want to check for indirect array accesses).

I think I would actually be fine to skip any non-loop code that is trailing (or leading) the loop nest. There isn't really any optimization opportunity for straight-line code.

huihzhang replied:
I agree on skipping any trailing or leading non-loop blocks.

Previously, I check for memory accesses, because non-loop blocks with no memory access doesn't cause much trouble
 for loop to be analyzed, although there is no use of adding these non-loop blocks either.

If non-loop blocks contain any memory accesses, yes, the address of these memory accesses would matter more than the value being loaded.

huihzhang committed:
Update region expansion profitability checks to reject any trailing non-loop blocks.

And address review comments.

huihzhang commented on polly/test/CodeGen/region_expansion_profitability_check.ll:
```diff
@@ -0,0 +1,138 @@
+; void test(int **restrict a, int *restrict b, int *restrict c,
+;           int *restrict d, int *restrict e, int *restrict f,
+;	    int L, int M, int Val0) {
+;   for (int i = 1; i <= L; i++) { // L1
```
I have simplified this test, `Loop i` and `Val0` are removed in the update.
I do need to keep `a` as multi-dimensional, as basePtr a[0] helps hit a scop invalidation on INVARIANTLOAD. And the scop is rejected unnecessarily.

huihzhang commented:
Thank you @Meinersbur for the review feedbacks!

I have pushed an update to check and reject if region expansion add any trailing non-loop blocks, with no following loops for potential loop fusion. Regardless of what these trailing non-loop blocks contain.

Please help take a look, let me know what you think? Thanks a lot!

/// a separate note ///
I will post a different patch to reject leading non-loop blocks. A quick try of this rejection below is affecting 141 polly unit tests.
```
@@ -1861,6 +1861,14 @@ bool ScopDetection::isValidRegion(DetectionContext &Context) {
     return false;
   }

+  // Reject if CurRegion has any leading non-loop blocks.
+  BasicBlock *Entry = CurRegion.getEntry();
+  Loop *L = LI.getLoopFor(Entry);
+  if (!L || !CurRegion.contains(L)) {
+    Context.IsInvalid = true;
+    return false;
+  }
+
   for (BasicBlock *Pred : predecessors(CurRegion.getEntry())) {
     Instruction *PredTerm = Pred->getTerminator();
     if (isa<IndirectBrInst>(PredTerm) || isa<CallBrInst>(PredTerm))
```

Meinersbur commented:
Your current approach is all-or-nothing, i.e. either use the expanded region or none at all. So I think the following is rejected:
```c
for (int i = 0; i < n; ++i)
  A[i] = i;
for (int j = 0; j < n; ++j)
  A[j] = j;
*unrelated = 42; 
```
That is, `expandRegion` is expanding to the j-loop (and remove the i-loop-only SCoP from the cache), then try to expand to `*unrelated` (and remove the i/j-loop-SCoP from the cache), then do your test which is failing and thus remove it from the cache as well. So at the end, not a single SCoP is accepted.

I think your test should be incorporated into the while-loop.

Your `isRegionExpansionProfitable` function can be massively simplified. No need to iterate over all BBs, just get the Exiting blocks and west whether they are in an included loop.
